src/main/java/com/google/sps/data/PersistentMatchRepository.java

```diff
@@ -0,0 +1,183 @@
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import com.google.sps.data.User;
+import java.io.IOException;
+import java.lang.IllegalArgumentException;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+/**
+ * This class saves User IDs of every User and their matches as Strings.
+ */
+public class PersistentMatchRepository implements MatchRepository {
+  
+  private final DatastoreService datastore;
+  private final Gson gson;
+  private final PersistentUserRepository userRepository;
+  
+  /**
+  * Constructor initializes private variables.
+  */
+  public PersistentMatchRepository() {
+      datastore = DatastoreServiceFactory.getDatastoreService();
+      gson = new Gson();
+      userRepository = new PersistentUserRepository();
+  }
+
+ /**
+  * Private method to add an entity to the datastore, or to override an existing 
+  * entity when new matches are added.
+  */
+  private void addNewEntity(String userId, String matchString) {
+    Entity userEntity = new Entity("User", userId);
+
+    userEntity.setProperty("userId", userId);
+    userEntity.setProperty("matchIds", matchString);
+    datastore.put(userEntity);
+  }
+
+ /**
+  * Private method to modify the json string and add a match to an existing user.
+  */
+  private void addMatchToExistingUser(String userId, String matchId) {
+    Collection<String> currentMatches = getMatchIdsForUser(userId);
+    currentMatches.add(matchId);
+    String matchString = gson.toJson(currentMatches);
+    addNewEntity(userId, matchString);
+  }
+
+  /**
+  * Private method that returns a Map of all the users and saved matches fitting 
+  * PreparedQuery results.
+  */
+  private Map<String, Collection<String>> fetchMapFromQuery(PreparedQuery results) {
+    Map<String, Collection<String>> idMap = new HashMap<String, Collection<String>>();
+
+    for (Entity entity : results.asIterable()) {
+        String userId = (String) entity.getProperty("userId");
+        String matchIds = (String) entity.getProperty("matchIds");
+        Collection<String> matchCollection = gson.fromJson(matchIds, HashSet.class);
+
+        idMap.put(userId, matchCollection);
+    }
+
+    return idMap;
+  }
+
+  /**
+  * Private method to fetch the full ID Map from the database.
+  */
+  private Map<String, Collection<String>> fetchFullMap() {
+    Query query = new Query("User");
+    PreparedQuery results = datastore.prepare(query);
+    Map<String, Collection<String>> idMap = fetchMapFromQuery(results);
+    return idMap;
+  }
+
+  /**
+  * Private method that sets a query filter based on User ID for the results in the datastore.
```

Comment:
supernit: You can probably omit things like "Private method that" or "class that" in favor of being more direct, e.g.:
`// Returns a datastore query filtered on 'userID' `. This is less proper grammatically, this is true, but more optimized for reading. Leaving this at your discretion since this is mostly personal style.

Reply:
Good idea, I changed them